doc/man/cgexec.1
.\" Copyright (C) 2009 Red Hat, Inc. All Rights Reserved.
.\" Written by Ivana Varekova <varekova@redhat.com>

.TH CGEXEC  1 2009-03-15 "Linux" "libcgroup Manual"
.SH NAME

cgexec \- run the task in given control groups

.SH SYNOPSIS
\fBcgexec\fR [\fB-h\fR] [\fB-b\fR] [\fB-g\fR <\fIcontrollers>:<path\fR>] [--sticky] \fBcommand\fR [\fIarguments\fR]

.SH DESCRIPTION
The \fBcgexec\fR
program executes the task \fBcommand\fR
with arguments \fBarguments\fR in the given control groups.

.TP
.B -b
ignores the default systemd delegated hierarchy path and
constructs the path of the control groups relative to the
cgroup root hierarchy.

.TP
.B -g <controllers>:<path>
defines the control groups in which the task will be run.
\fBcontrollers\fR is a list of controllers and
\fBpath\fR is the relative path to control groups
in the given controllers list.

This flag can be used multiple times to
define multiple pairs of lists of controllers
and relative paths.
Instead of the list of all mounted controllers,
the wildcard \fBb"*b"\fR can be used.

If this option is not used,
\fBcgexec\fR will automatically place the task in the right
cgroup based on \fB/etc/cgrules.conf\fR.

If \fB/etc/cgrules.conf\fR configuration file is used, there can be used
template names. Then the control group name contains
a template in destination tag (see \fBcgrules.conf\fR (5))
and if the cgroup does not exist in execution time,
it is created, based on \fB/etc/cgconfig.conf\fR specification.
If the specifications are not present the group is created with the default
kernel values.

.TP
.B -h, --help
Display this help and exit.

.TP
.B --sticky
If running the task \fBcommand\fR with this option, the daemon of
service cgred (cgrulesengd process) does not change both the task
of the \fBcommand\fR and the child tasks. Without this option,
the daemon does not change the task of the \fBcommand\fR but it
changes the child tasks to the right cgroup based on
\fB/etc/cgrules.conf\fR automatically.

.LP

.SH EXAMPLES

.nf
.ft B
cgexec -g *:test1 ls
.ft R
.fi
runs command \fBls\fR in control group test1
in all mounted controllers.

.nf
.ft B
cgexec -g cpu,memory:test1 ls -l
.ft R
.fi
runs command \fBls -l\fR in control group test1
in controllers cpu and memory.

.nf
.ft B
cgexec -g cpu,memory:test1 -g swap:test2 ls -l
.ft R
.fi
runs command \fBls -l\fR in control group test1
in controllers cpu and memory and control group
test2 in controller swap.

.SH ENVIRONMENT VARIABLES
.TP
.B CGROUP_LOGLEVEL
controls verbosity of the tool. Allowed values are \fBDEBUG\fR,
\fBINFO\fR, \fBWARNING\fR or \fBERROR\fR.

.SH FILES
.TP
.B /etc/cgrules.conf
default libcgroup configuration file

.SH SEE ALSO
cgrules.conf (5)


src/tools/cgexec.c
// SPDX-License-Identifier: LGPL-2.1-only
/**
 * Copyright RedHat Inc. 2008
 *
 * Authors:	Vivek Goyal <vgoyal@redhat.com>
 */

#ifndef _GNU_SOURCE
#define _GNU_SOURCE
#endif

#include "tools-common.h"

#include <libcgroup.h>

#include <limits.h>
#include <search.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <getopt.h>
#include <stdio.h>
#include <errno.h>
#include <grp.h>
#include <pwd.h>

#include <sys/mount.h>
#include <sys/stat.h>
#include <sys/types.h>


static struct option longopts[] = {
	{"sticky",	no_argument, NULL, 's'},
	{"help",	no_argument, NULL, 'h'},
	{0, 0, 0, 0}
};

static void usage(int status, const char *program_name)
{
	if (status != 0) {
		err("Wrong input parameters, try %s --help for more information.\n", program_name);
		return;
	}

	info("Usage: %s [-h] [-g <controllers>:<path>] [--sticky] ", program_name);
	info("command [arguments] ...\n");
	info("Run the task in given control group(s)\n");
	info("  -g <controllers>:<path>	Control group which should be added\n");
	info("  -h, --help			Display this help\n");
	info("  --sticky			cgred daemon does not ");
	info("change pidlist and children tasks\n");
#ifdef WITH_SYSTEMD
	info("  -b				Ignore default systemd delegate hierarchy\n");
#endif
}

int main(int argc, char *argv[])
{
	struct cgroup_group_spec *cgroup_list[CG_HIER_MAX];
	int ignore_default_systemd_delegate_slice = 0;
	int cg_specified = 0;
	int flag_child = 0;
	int i, ret = 0;
	uid_t uid;
	gid_t gid;
	pid_t pid;
	int c;

	memset(cgroup_list, 0, sizeof(cgroup_list));
#ifdef WITH_SYSTEMD
	while ((c = getopt_long(argc, argv, "+g:shb", longopts, NULL)) > 0) {
		switch (c) {
		case 'b':
			ignore_default_systemd_delegate_slice = 1;
			break;
#else
	while ((c = getopt_long(argc, argv, "+g:sh", longopts, NULL)) > 0) {
		switch (c) {
#endif
		case 'g':
			ret = parse_cgroup_spec(cgroup_list, optarg, CG_HIER_MAX);
			if (ret) {
				err("cgroup controller and path parsing failed\n");
				exit(EXIT_BADARGS);
			}
			cg_specified = 1;
			break;
		case 's':
			flag_child |= CGROUP_DAEMON_UNCHANGE_CHILDREN;
			break;
		case 'h':
			usage(0, argv[0]);
			exit(0);
		default:
			usage(1, argv[0]);
			exit(EXIT_BADARGS);
		}
	}

	/* Executable name */
	if (!argv[optind]) {
		usage(1, argv[0]);
		exit(EXIT_BADARGS);
	}

	/* Initialize libcg */
	ret = cgroup_init();
	if (ret) {
		err("libcgroup initialization failed: %s\n", cgroup_strerror(ret));
		return ret;
	}

	/* this is false always for disable-systemd */
	if (!ignore_default_systemd_delegate_slice)
		cgroup_set_default_systemd_cgroup();

	/* Just for debugging purposes. */
	uid = geteuid();
	gid = getegid();
	cgroup_dbg("My euid and egid is: %d,%d\n", (int) uid, (int) gid);

	uid = getuid();
	gid = getgid();
	pid = getpid();

	ret = cgroup_register_unchanged_process(pid, flag_child);
	if (ret) {
		err("registration of process failed\n");
		return ret;
	}

	/*
	 * 'cgexec' command file needs the root privilege for executing a
	 * cgroup_register_unchanged_process() by using unix domain socket,
	 * and an euid/egid should be changed to the executing user from a
	 * root user.
	 */
	if (setresuid(uid, uid, uid)) {
		err("%s", strerror(errno));
		return -1;
	}

	if (setresgid(gid, gid, gid)) {
		err("%s", strerror(errno));
		return -1;
	}

	if (cg_specified) {
		/*
		 * User has specified the list of control group
		 * and controllers
		 */
		for (i = 0; i < CG_HIER_MAX; i++) {
			if (!cgroup_list[i])
				break;

			ret = cgroup_change_cgroup_path(cgroup_list[i]->path, pid,
				(const char *const*) cgroup_list[i]->controllers);
			if (ret) {
				err("cgroup change of group failed\n");
				return ret;
			}
		}
	} else {

		/* Change the cgroup by determining the rules based on uid */
		ret = cgroup_change_cgroup_flags(uid, gid, argv[optind], pid, 0);
		if (ret) {
			err("cgroup change of group failed\n");
			return ret;
		}
	}

	/* Now exec the new process */
	ret = execvp(argv[optind], &argv[optind]);
	if (ret == -1) {
		err("%s", strerror(errno));
		return -1;
	}

	return 0;
}
